types/time.go

```diff
@@ -207,159 +184,202 @@ func FromGoTime(t gotime.Time) MysqlTime {
 }
 
 // FromDate makes a internal time representation from the given date.
-func FromDate(year int, month int, day int, hour int, minute int, second int, microsecond int) MysqlTime {
-	return MysqlTime{
-		year:        uint16(year),
-		month:       uint8(month),
-		day:         uint8(day),
-		hour:        uint32(hour),
-		minute:      uint8(minute),
-		second:      uint8(second),
-		microsecond: uint32(microsecond),
-	}
+func FromDate(year int, month int, day int, hour int, minute int, second int, microsecond int) CoreTime {
+	t := ZeroCoreTime
+	p := (*uint64)(&t)
+	*p |= (uint64(microsecond) << microsecondBitFieldOffset) & microsecondBitFieldMask
+	*p |= (uint64(second) << secondBitFieldOffset) & secondBitFieldMask
+	*p |= (uint64(minute) << minuteBitFieldOffset) & minuteBitFieldMask
+	*p |= (uint64(hour) << hourBitFieldOffset) & hourBitFieldMask
+	*p |= (uint64(day) << dayBitFieldOffset) & dayBitFieldMask
+	*p |= (uint64(month) << monthBitFieldOffset) & monthBitFieldMask
+	*p |= (uint64(year) << yearBitFieldOffset) & yearBitFieldMask
+	return t
 }
 
-// MaxMySQLTime returns Duration with maximum mysql time.
-func MaxMySQLTime(fsp int8) Time {
-	return NewTime(FromDate(0, 0, 0, TimeMaxHour, TimeMaxMinute, TimeMaxSecond, 0), mysql.TypeDuration, fsp)
+// FromDateChecked makes a internal time representation from the given date with field overflow check.
+func FromDateChecked(year, month, day, hour, minute, second, microsecond int) (CoreTime, bool) {
+	if uint64(year) >= (1 << yearBitFieldWidth) {
+		return ZeroCoreTime, false
+	}
+	if uint64(month) >= (1 << monthBitFieldWidth) {
+		return ZeroCoreTime, false
+	}
+	if uint64(day) >= (1 << dayBitFieldWidth) {
+		return ZeroCoreTime, false
+	}
+	if uint64(hour) >= (1 << hourBitFieldWidth) {
+		return ZeroCoreTime, false
+	}
+	if uint64(minute) >= (1 << minuteBitFieldWidth) {
+		return ZeroCoreTime, false
+	}
+	if uint64(second) >= (1 << secondBitFieldWidth) {
+		return ZeroCoreTime, false
+	}
+	if uint64(microsecond) >= (1 << microsecondBitFieldWidth) {
+		return ZeroCoreTime, false
+	}
+	return FromDate(year, month, day, hour, minute, second, microsecond), true
 }
 
-// Clock returns the hour, minute, and second within the day specified by t.
-func (t Time) Clock() (hour int, minute int, second int) {
-	return t.time.Hour(), t.time.Minute(), t.time.Second()
-}
+// coreTime is an alias to CoreTime, embedd in Time.
+type coreTime = CoreTime
 
 // Time is the struct for handling datetime, timestamp and date.
-// TODO: check if need a NewTime function to set Fsp default value?
 type Time struct {
-	time MysqlTime
-	tp   uint8
-	// Fsp is short for Fractional Seconds Precision.
-	// See http://dev.mysql.com/doc/refman/5.7/en/fractional-seconds.html
-	fsp int8
+	coreTime
```

Comment:
can we move the functions of `coreTime` to `Time` and remove the definition of `coreTime`?

Reply:
I merge CoreTime and Time at first, but @breeswish say that it's dangerous if code in other module call CoreTime() then call type related function on it, so I remains the CoreTime struct.

Reply:
@zz-jason `CoreTime` and `Time` are different types (CoreTime is YMDHMS part, while Time contains CoreTime + fsp + type). I am worried about making mistakes if CoreTime and Time use same type (mixed use will not be detected by the compiler).